docs: Port "Integrations" extension page to 7.2 (#590)
mautic/developer-documentation-new: docs/plugin_extensions/integrations.rst

   Please read the :xref:`dev docs contributing guidelines` and :xref:`Contributing to Mautic’s documentation` to get started.

> .. vale on

Reviewer: Remove the note.

Integrate third-party services with Mautic by defining an Integration class for the service. Each Integration class handles the authorization process, configuration, and requests for that service.

.. note::

>    This page covers the legacy ``AbstractIntegration`` approach. New Integrations should use the :doc:`Integrations framework</plugin_integrations/integrations>`, which separates authentication, configuration, and sync concerns into dedicated interfaces.

Author: Issue #328 (parent of #329) supplied the legacy source URL and the "transfer the content under the note admonition" guidance, so the ported content is placed under the existing `.. note::` placeholder and the legacy `AbstractIntegration` approach is cross-linked to the modern `plugin_integrations/integrations` framework page.

Source: https://github.com/mautic/developer-documentation-new/issues/328

>    This page covers the legacy ``AbstractIntegration`` approach. New Integrations should use the :doc:`Integrations framework</plugin_integrations/integrations>`, which separates authentication, configuration, and sync concerns into dedicated interfaces.

Reviewer: Wrap this note admonition in vale off/on statements.

Integration class
*****************

> A Plugin can provide multiple Integrations, each defined as its own class in the bundle's ``Integration`` folder. The class extends ``Mautic\PluginBundle\Integration\AbstractIntegration``, which provides many helper methods, including OAuth authorization and request signing:

Author: Ported the `Integration class`, `Integration image`, `Authorization`, `Functions`, and `makeRequest` sections plus the `MarsIntegration` example from the legacy `_plugin_extending_integrations.md` source, cleaning up the duplicated `<?php` line and the missing semicolon after `return 'Red Mars'` and adding PHP return types per the org style guide.

Source: https://github.com/mautic/developer-documentation/blob/main/source/includes/_plugin_extending_integrations.md

.. code-block:: php

    <?php
    // plugins/HelloWorldBundle/Integration/MarsIntegration.php

    namespace MauticPlugin\HelloWorldBundle\Integration;

    use Mautic\PluginBundle\Integration\AbstractIntegration;

    class MarsIntegration extends AbstractIntegration
    {
        public function getName(): string
        {
            return 'Mars';
        }

        public function getDisplayName(): string
        {
            return 'Red Mars';
        }

        public function getAuthenticationType(): string
        {
            return 'oauth2';
        }

        public function getAuthenticationUrl(): string
        {
            return 'https://example.com/oauth/authorize';
        }

        public function getAccessTokenUrl(): string
        {
            return 'https://example.com/oauth/access_token';
        }

        public function getSupportedFeatures(): array
        {
            return ['public_profile', 'public_activity'];
        }
    }

``getName()`` must match the file - for example, ``IcontactIntegration`` returns ``Icontact``. ``getDisplayName()`` defaults to ``getName()`` unless you override it.

Integration image
*****************

> Each Integration is displayed on a card in the 'Manage Plugins' area. To set its image, add a 128x128px PNG to the bundle's ``Assets/img`` folder, named after the value returned by ``getName()`` in lowercase. For example, ``MarsIntegration`` uses ``plugins/HelloWorldBundle/Assets/img/mars.png``.

Reviewer: There are a few things here:

- Add PNG to accept.txt file
- Address Vale warning:  [Google.Passive] In general, use active voice instead of passive voice ('is displayed').
- Add a space for '128x128 px'

```suggestion
Each Integration is displayed on a card in the 'Manage Plugins' area. To set its image, add a 128x128 px PNG to the bundle's ``Assets/img`` folder, named after the value returned by ``getName()`` in lowercase. For example, ``MarsIntegration`` uses ``plugins/HelloWorldBundle/Assets/img/mars.png``.
```

Authorization
*************

> Out of the box, ``AbstractIntegration`` handles the standard key, OAuth1a, and OAuth2 specifications. The ``getAuthenticationType()`` method defines which one to use. Each input the User must provide - username, password, and so on - is defined by an array of ``keyName => label`` elements returned by ``getRequiredKeyFields()``. That method isn't required when you use one of the standard specifications.

Reviewer: Use active instead of passive voice. See: 'is defined' and 'isn't required'

> Keys saved by the Integration are encrypted. To access the decrypted values inside the Integration class, use the ``$this->keys`` array. You can override any method defined in ``AbstractIntegration`` to suit the Integration's needs.

Reviewer: Address the Vale warning: 

- [Google.Passive] In general, use active voice instead of passive voice ('are encrypted').

Functions
*********

> Some of the main methods are described below. Review the ``AbstractIntegration`` class and its method doc blocks for more detail.

Reviewer: Address the Vale warning:

- [Google.Passive] In general, use active voice instead of passive voice ('are described').

.. list-table::
   :header-rows: 1

   * - Area
     - Function
     - Description
   * - Auth
     - ``getRequiredKeyFields``

>      - Returns an array of ``keyName => label`` elements for settings the User must provide, such as username, password, client ID, or client secret. Each element is displayed as an input in the Integration's settings.

Reviewer: Address the Vale warning:

- [Google.Passive] In general, use active voice instead of passive voice ('is displayed').

   * - Auth
     - ``getSecretKeys``

>      - Returns any ``keyName`` from ``getRequiredKeyFields`` that is secret, so it's masked in the form.

Reviewer: There are a few things here:

- Use active instead of passive voice. See: 'it's masked'

- Address the Vale warnings:

  - [Mautic.FeatureList] Is this referring to a Mautic feature? If so, use 'Form' instead of 'form'.
  - [Google.Contractions] Feel free to use 'that's' instead of 'that is'.

   * - Auth & Request
     - ``getClientIdKey``

>      - Defines the "username" for the Integration. Defaults to ``client_id`` for OAuth2 and ``keyName`` for the key authentication type.

Reviewer: ```suggestion
     - Defines the 'username' for the Integration. Defaults to ``client_id`` for OAuth2 and ``keyName`` for the key authentication type.
```

   * - Auth & Request
     - ``getClientSecretKey``

>      - Defines the "password" for the Integration. By default, only OAuth2 uses this and returns ``client_secret``.

Reviewer: ```suggestion
     - Defines the 'password' for the Integration. By default, only OAuth2 uses this and returns ``client_secret``.
```

   * - Auth
     - ``getAuthLoginUrl``
     - Defines the login URL for the OAuth1a specification.
   * - Auth
     - ``getAuthenticationUrl``
     - Defines the login URL for the OAuth2 specification.
   * - Auth
     - ``getAccessTokenUrl``
     - Defines the access token URL for the OAuth2 specification.
   * - Auth
     - ``getAuthScope``
     - Defines the scope for the OAuth2 specification.
   * - Auth
     - ``getAuthCallbackUrl``
     - Defines the callback URL for the OAuth1a or OAuth2 specifications. Defaults to the ``mautic_integration_auth_callback`` route.
   * - General
     - ``isConfigured``
     - Determines whether the Integration has been configured correctly.
   * - General
     - ``isAuthorized``

>      - Determines whether the Integration is authorized, reauthorizing when an OAuth2 refresh token is stored.

Reviewer: Use active instead of passive voice. See: 'is authorized' and 'is stored'.

   * - Request
     - ``makeRequest``
     - Makes API requests, automatically handling the standard key, OAuth1a, and OAuth2 specifications.
   * - Form
     - ``getFormSettings``
     - Returns options for the Integration's configuration form. Supported keys are ``requires_callback`` and ``requires_authorization``.
   * - Form
     - ``getFormNotes``
     - Returns helper notes to display in the form.

.. vale off

makeRequest
===========

.. vale on

``makeRequest()`` can automatically sign outgoing requests and handle authentication. Any Integration can override it. It accepts the following parameters:

.. list-table::
   :header-rows: 1

   * - Name
     - Type
     - Description
   * - ``$url``
     - string
     - The URL to make the request to.
   * - ``$parameters``
     - array

>      - Parameters to submit with the request. For a ``GET`` request they're appended to the query string; otherwise they form the POST body.

Reviewer: Use active instead of passive voice. See: 'they're appended'

   * - ``$method``
     - string
     - The request method - ``GET``, ``POST``, ``PUT``, ``PATCH``, or ``DELETE``.
   * - ``$settings``
     - array

>      - Configures the behavior of ``makeRequest()``. Built-in optional settings are listed below.

Reviewer: Use active instead of passive voice. See: 'are listed'.
